rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/SchedulerStateRest.java

```diff
@@ -1015,24 +1038,64 @@ private String getTaskResultValueAsStringOrExceptionStackTrace(TaskResult taskRe
     }
 
     @Override
-    public byte[] serializedValueOfTaskResult(String sessionId, String jobId, String taskname) throws Throwable {
+    public byte[] serializedValueOfTaskResult(String sessionId, String jobId, String taskname) throws RestException {
         Scheduler s = checkAccess(sessionId, PATH_JOBS + jobId + PATH_TASKS + taskname + "/result/serializedvalue");
-        TaskResult tr = s.getTaskResult(jobId, taskname);
-        tr = PAFuture.getFutureValue(tr);
-        return tr.getSerializedValue();
+        try {
+            TaskResult tr = s.getTaskResult(jobId, taskname);
+            tr = PAFuture.getFutureValue(tr);
+            return tr.getSerializedValue();
+        } catch (SchedulerException e) {
+            throw RestException.wrapExceptionToRest(e);
+        }
+    }
+
+    @Override
+    public Response downloadTaskResult(String sessionId, String jobId, String taskname, String destination)
+            throws RestException {
+        byte[] serializedValue = serializedValueOfTaskResult(sessionId, jobId, taskname);
+        Map<String, String> resultMetadata = metadataOfTaskResult(sessionId, jobId, taskname);
+        if (resultMetadata == null) {
+            resultMetadata = new HashMap<>();
+        }
+        if (destination == null) {
+            destination = DESTINATION_BROWSER;
+        }
+
+        Response.ResponseBuilder builder = Response.ok().entity(new ByteArrayInputStream(serializedValue));
+
+        if (resultMetadata.containsKey(METADATA_CONTENT_TYPE)) {
+            builder = builder.header(HttpHeaders.CONTENT_TYPE, resultMetadata.get(METADATA_CONTENT_TYPE));
+        }
+        if (DESTINATION_FILE.equals(destination)) {
+            if (resultMetadata.containsKey(METADATA_FILE_NAME)) {
+                builder = builder.header(HttpHeaders.CONTENT_DISPOSITION,
+                                         "attachment; filename=" + resultMetadata.get(METADATA_FILE_NAME));
+            } else if (resultMetadata.containsKey((METADATA_FILE_EXTENSION))) {
```

Comment:
what if both METADATA_FILE_NAME &  METADATA_FILE_EXTENSION are provided in the following fashion :
fileName = "myFile", extension = "pdf" ?
In this case, you'll download a file called myFile without the proper extension. If it's an executable or any other type of file without a standard content-type then users will have to rename the file after it's downloaded to get the extension right.

Maybe you can simply do :

```
if (DESTINATION_FILE.equals(destination)) {
    String fileExtension= resultMetadata.containsKey(METADATA_FILE_EXTENSION)?resultMetadata.get(METADATA_FILE_EXTENSION):"";
    String fileName = resultMetadata.containsKey(METADATA_FILE_NAME)?resultMetadata.get(METADATA_FILE_NAME):"job_" + jobId + "_" + taskname + "_result";
    builder = builder.header(HttpHeaders.CONTENT_DISPOSITION,
                                          "attachment; filename= filename+fileExtension;
         }
```

Reply:
METADATA_FILE_NAME takes precedence (it must include the extension)

Reply:
Maybe we can check when both are present but the filename does or doesn't contain an extension ? Just in case resultMetadata is misused.
Something like 'if (fileName.endsWith(fileExtension))' ?

Reply:
It is explained in the documentation
https://doc.activeeon.com/latest/user/ProActiveUserGuide.html#_assigning_metadata_to_task_result

I am not going to handle misuse.